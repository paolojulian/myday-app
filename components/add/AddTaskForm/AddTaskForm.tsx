import { TabName } from '@/app/(tabs)/_layout';
import {
  ADD_TASK_VALIDATION_SCHEMA,
  convertTaskFormToTask,
  NOTE_PLACEHOLDER,
  TaskFormValues,
} from '@/components/add/AddTaskForm/AddTaskForm.utils';
import Button from '@/components/common/Button';
import Container from '@/components/common/Container';
import EasyDatePicker from '@/components/common/EasyDatePicker';
import CheckboxField from '@/components/common/forms/CheckboxField';
import TextArea from '@/components/common/forms/TextArea';
import TextField from '@/components/common/forms/TextField';
import Snackbar from '@/components/common/Snackbar';
import ThemedView from '@/components/common/ThemedView';
import { useCreateTask } from '@/hooks/services/task/useCreateTask';
import { GlobalSnackbar } from '@/managers/SnackbarManager';
import { useNavigation } from 'expo-router';
import { Formik } from 'formik';
import { Fragment, useRef, useState } from 'react';
import { ScrollView, TextInput } from 'react-native';

export const ADD_TASK_FORM_TEST_IDS = {
  title: 'add-task-form-title',
};

export default function AddTaskForm() {
  const noteRef = useRef<TextInput>(null);
  const amountRef = useRef<TextInput>(null);
  const navigation = useNavigation();
  const [error, setError] = useState<string | null>(null);
  const { mutate: createTaskMutate } = useCreateTask();

  const showSuccessMessage = () => {
    GlobalSnackbar.show({
      message: 'Todo created successfully',
      duration: GlobalSnackbar.LENGTH_LONG,
      type: 'success',
    });
  };

  const handleSubmitForm = async (values: TaskFormValues) => {
    const insertValues = convertTaskFormToTask(values);
    try {
      await createTaskMutate(insertValues);
      showSuccessMessage();
      navigation.navigate(TabName.Todo as never);
    } catch {
      setError('Failed to create expense');
    }
  };

  const handleTitleSubmitEditing = () => {
    noteRef.current?.focus();
  };
  const handleToBuyChecked = (value: boolean) => {
    if (value) {
      setTimeout(() => {
        amountRef.current?.focus();
      }, 0);
    }
  };

  return (
    <Fragment>
      <ScrollView style={{ flex: 1 }} keyboardShouldPersistTaps="handled" directionalLockEnabled>
        <Container style={{ flex: 1 }}>
          <Formik<TaskFormValues>
            onSubmit={handleSubmitForm}
            initialValues={{
              title: '',
              category: '',
              description: '',
              toBuy: false,
            }}
            validationSchema={ADD_TASK_VALIDATION_SCHEMA}
          >
            {({
              handleSubmit,
              handleChange,
              handleBlur,
              setFieldValue,
              values,
              touched,
              errors,
            }) => (
              <>
                <ThemedView style={{ gap: 8, flex: 1 }}>
                  <TextField
                    onSubmitEditing={handleTitleSubmitEditing}
                    onChangeText={handleChange('title')}
                    onBlur={handleBlur('title')}
                    testID={ADD_TASK_FORM_TEST_IDS.title}
                    isError={touched.title && !!errors.title}
                    errorMessage={errors.title}
                    autoFocus
                    value={values.title}
                    label="Title"
                    placeholder="Buy groceries..."
                    returnKeyLabel="Next"
                    returnKeyType="next"
                  />
                  <TextArea
                    onChangeText={handleChange('description')}
                    onBlur={handleBlur('description')}
                    ref={noteRef}
                    label="Note (Optional)"
                    placeholder={NOTE_PLACEHOLDER}
                    isError={touched.description && !!errors.description}
                    errorMessage={errors.description}
                    value={values.description}
                  />
                  <EasyDatePicker
                    onSelectDate={date => setFieldValue('reminderDate', date)}
                    selectedDate={values.reminderDate}
                    label="Reminder Date"
                  />
                  <CheckboxField
                    onValueChange={value => {
                      setFieldValue('toBuy', value);
                      if (!value) setFieldValue('amount', null);
                      handleToBuyChecked(value);
                    }}
                    label={'To Buy'}
                    value={!!values.toBuy}
                  />
                  {values.toBuy && (
                    <TextField
                      ref={amountRef}
                      label="Expected amount (Optional)"
                      placeholder="00.00"
                      keyboardType="numeric"
                      value={values.amount}
                      onChangeText={handleChange('amount')}
                      returnKeyType="done"
                    />
                  )}
                </ThemedView>
                <ThemedView style={{ marginVertical: 16 }}>
                  <Button text={'Save'} onPress={() => handleSubmit()} />
                </ThemedView>
              </>
            )}
          </Formik>
        </Container>
      </ScrollView>

      {/* Error snackbar */}
      <Snackbar onDismiss={() => setError('')} message={error} type="error" />
    </Fragment>
  );
}
