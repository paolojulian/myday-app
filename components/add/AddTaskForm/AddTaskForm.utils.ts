import { SupportedCreateTaskFields } from '@/hooks/services/task/useCreateTask';
import { convertDateToEpoch } from '@/utils/date/date.utils';
import * as Yup from 'yup';

export const ADD_TASK_VALIDATION_SCHEMA = Yup.object().shape({
  title: Yup.string().max(40).required('Title is required'),
  description: Yup.string().max(255),
  category: Yup.string(),
  reminderDate: Yup.date().optional(),
  toBuy: Yup.boolean().optional(),
  amount: Yup.string().optional(),
});
export type TaskFormValues = Yup.InferType<typeof ADD_TASK_VALIDATION_SCHEMA>;

export function convertTaskFormToTask(taskForm: TaskFormValues): SupportedCreateTaskFields {
  return {
    title: taskForm.title,
    description: taskForm.description ?? '',
    is_completed: 0,
    to_buy: taskForm.toBuy ? 1 : 0,
    expected_amount: taskForm.amount ? parseFloat(taskForm.amount) : null,
    reminder_date: taskForm.reminderDate ? convertDateToEpoch(taskForm.reminderDate) : null,
  };
}

export const NOTE_PLACEHOLDER = `e.g.
Dish soap
Soy sauce
Toilet paper
etc..`;
