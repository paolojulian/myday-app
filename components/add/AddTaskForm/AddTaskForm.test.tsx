import { act, fireEvent, render, screen } from '@testing-library/react-native';
import AddTaskForm from './AddTaskForm';

jest.mock('expo-router', () => ({
  ...jest.requireActual('expo-router'),
  useNavigation: jest.fn(),
}));

jest.mock('@/hooks/services/task/useCreateTask', () => ({
  useCreateTask: jest.fn(() => ({
    mutate: jest.fn(),
  })),
}));

describe('TESTING AddTaskForm', () => {
  describe('GIVEN the AddTaskForm component', () => {
    describe('WHEN it is rendered', () => {
      it('THEN should match the snapshot', () => {
        const component = renderComponent();
        expect(component).toMatchSnapshot();
      });
      it('THEN it should display the form', () => {
        renderComponent();

        const title = screen.getByLabelText('Title');
        expect(title).toBeDefined();

        const description = screen.getByLabelText('Note (Optional)');
        expect(description).toBeDefined();

        const reminderDate = screen.getByLabelText('Reminder Date (Optional)');
        expect(reminderDate).toBeDefined();

        expect(screen.getByLabelText('To Buy')).toBeDefined();

        act(() => {
          fireEvent(screen.getByLabelText('To Buy'), 'press');
        });

        const amount = screen.getByLabelText('Expected amount (Optional)');
        expect(amount).toBeDefined();
      });
    });
  });
});

function renderComponent() {
  return render(<AddTaskForm />);
}
