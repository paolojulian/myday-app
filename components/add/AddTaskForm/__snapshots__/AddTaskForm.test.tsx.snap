// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TESTING AddTaskForm GIVEN the AddTaskForm component WHEN it is rendered THEN should match the snapshot 1`] = `
[
  <RCTScrollView
    directionalLockEnabled={true}
    keyboardShouldPersistTaps="handled"
    style={
      {
        "flex": 1,
      }
    }
  >
    <View>
      <View
        style={
          [
            {
              "paddingHorizontal": 16,
            },
            {
              "flex": 1,
            },
          ]
        }
      >
        <View
          style={
            {
              "flex": 1,
              "gap": 8,
            }
          }
        >
          <View
            style={
              [
                {
                  "position": "relative",
                },
              ]
            }
          >
            <View
              style={
                {
                  "left": 16,
                  "position": "absolute",
                  "top": 12,
                  "zIndex": 1,
                }
              }
            >
              <Text
                aria-label="Title"
                style={
                  [
                    {
                      "fontFamily": "PoppinsMedium",
                      "fontSize": 16,
                      "fontWeight": 500,
                    },
                    {
                      "fontWeight": 400,
                    },
                    {
                      "color": "#0F172A",
                    },
                  ]
                }
              >
                Title
              </Text>
            </View>
            <TextInput
              autoFocus={true}
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              onSubmitEditing={[Function]}
              placeholder="Buy groceries..."
              placeholderTextColor="#a3a3a3"
              returnKeyLabel="Next"
              returnKeyType="next"
              style={
                [
                  undefined,
                  {
                    "backgroundColor": "#f5f5f5",
                    "borderRadius": 8,
                    "borderWidth": 1,
                    "paddingBottom": 12,
                    "paddingHorizontal": 16,
                    "paddingTop": 36,
                  },
                  {
                    "fontFamily": "PoppinsRegular",
                    "fontSize": 16,
                  },
                  {
                    "borderColor": "transparent",
                  },
                ]
              }
              testID="add-task-form-title"
              value=""
            />
          </View>
          <View
            style={
              [
                {
                  "position": "relative",
                },
              ]
            }
          >
            <View
              style={
                {
                  "left": 16,
                  "position": "absolute",
                  "top": 12,
                  "zIndex": 1,
                }
              }
            >
              <Text
                aria-label="Note (Optional)"
                style={
                  [
                    {
                      "fontFamily": "PoppinsMedium",
                      "fontSize": 16,
                      "fontWeight": 500,
                    },
                    {
                      "fontWeight": 400,
                    },
                    {
                      "color": "#0F172A",
                    },
                  ]
                }
              >
                Note (Optional)
              </Text>
            </View>
            <TextInput
              multiline={true}
              numberOfLines={4}
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              placeholder="e.g.
Dish soap
Soy sauce
Toilet paper
etc.."
              placeholderTextColor="#a3a3a3"
              style={
                [
                  {
                    "minHeight": 120,
                  },
                  {
                    "backgroundColor": "#f5f5f5",
                    "borderRadius": 8,
                    "borderWidth": 1,
                    "paddingBottom": 12,
                    "paddingHorizontal": 16,
                    "paddingTop": 36,
                  },
                  {
                    "fontFamily": "PoppinsRegular",
                    "fontSize": 16,
                  },
                  {
                    "borderColor": "transparent",
                  },
                ]
              }
              testID="textField__textInput"
              value=""
            />
          </View>
          <View
            style={
              [
                {},
                {
                  "borderColor": "#f5f5f5",
                  "borderRadius": 12,
                  "borderWidth": 2,
                },
              ]
            }
            testID="date-picker-container"
          >
            <View
              style={
                {
                  "alignItems": "flex-end",
                  "backgroundColor": "#f5f5f5",
                  "borderTopEndRadius": 12,
                  "borderTopStartRadius": 12,
                  "display": "flex",
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "paddingHorizontal": 16,
                  "paddingVertical": 16,
                }
              }
            >
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "alignItems": "center",
                    "flex": 1,
                    "flexDirection": "row",
                    "gap": 16,
                  }
                }
                testID="date-picker-header-btn"
              >
                <Image
                  source={1}
                  style={
                    {
                      "height": 24,
                      "width": 24,
                    }
                  }
                />
                <View
                  style={
                    {
                      "flex": 1,
                      "gap": 4,
                    }
                  }
                >
                  <Text
                    aria-label="Reminder Date (Optional)"
                    style={
                      [
                        {
                          "fontFamily": "PoppinsMedium",
                          "fontSize": 16,
                          "fontWeight": 500,
                        },
                        {
                          "fontWeight": 400,
                        },
                        {
                          "color": "#0F172A",
                        },
                      ]
                    }
                  >
                    Reminder Date (Optional)
                  </Text>
                  <Text
                    style={
                      [
                        {
                          "fontFamily": "PoppinsRegular",
                          "fontSize": 16,
                        },
                        {
                          "fontWeight": 400,
                        },
                        {
                          "color": "#a3a3a3",
                        },
                      ]
                    }
                    testID="date-picker-value"
                  >
                    Select a date
                  </Text>
                </View>
              </View>
            </View>
          </View>
          <RNGestureHandlerButton
            collapsable={false}
            delayLongPress={600}
            enabled={true}
            exclusive={true}
            handlerTag={1}
            handlerType="NativeViewGestureHandler"
            onGestureEvent={[Function]}
            onGestureHandlerEvent={[Function]}
            onGestureHandlerStateChange={[Function]}
            onHandlerStateChange={[Function]}
            rippleColor={0}
            touchSoundDisabled={false}
          >
            <View
              accessible={true}
              collapsable={false}
              style={
                {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "#f5f5f5",
                    "flexDirection": "row",
                    "gap": 8,
                    "paddingHorizontal": 24,
                    "paddingVertical": 16,
                  }
                }
              >
                <View
                  accessibilityRole="checkbox"
                  accessibilityState={
                    {
                      "busy": undefined,
                      "checked": false,
                      "disabled": undefined,
                      "expanded": undefined,
                      "selected": undefined,
                    }
                  }
                  accessibilityValue={
                    {
                      "max": undefined,
                      "min": undefined,
                      "now": undefined,
                      "text": undefined,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    [
                      {
                        "borderColor": "#657786",
                        "borderRadius": 2,
                        "borderWidth": 2,
                        "height": 20,
                        "width": 20,
                      },
                      {
                        "borderRadius": 4,
                      },
                      false,
                      {
                        "backgroundColor": undefined,
                        "borderColor": "#0F172A",
                      },
                      undefined,
                      false,
                    ]
                  }
                />
                <Text
                  aria-label="To Buy"
                  style={
                    [
                      {
                        "fontFamily": "PoppinsRegular",
                        "fontSize": 16,
                      },
                      {
                        "fontWeight": 400,
                      },
                      undefined,
                    ]
                  }
                >
                  To Buy
                </Text>
              </View>
            </View>
          </RNGestureHandlerButton>
        </View>
        <View
          style={
            {
              "marginVertical": 16,
            }
          }
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "borderRadius": 100,
                    "justifyContent": "center",
                    "padding": 16,
                  },
                  {
                    "backgroundColor": "#0F172A",
                  },
                ]
              }
            >
              <Text
                style={
                  [
                    {
                      "fontFamily": "PoppinsMedium",
                      "fontSize": 16,
                      "fontWeight": 500,
                    },
                    {
                      "fontWeight": 400,
                    },
                    [
                      {
                        "color": "#FFFFFF",
                      },
                      {
                        "fontWeight": "bold",
                      },
                    ],
                  ]
                }
              >
                Save
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>,
  <View
    collapsable={false}
    style={
      {
        "bottom": 0,
        "flex": 1,
        "left": 0,
        "marginBottom": 120,
        "opacity": 0,
        "pointerEvents": "none",
        "position": "absolute",
        "right": 0,
        "zIndex": 999,
      }
    }
  >
    <View
      collapsable={false}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#CC3333",
          "borderRadius": 8,
          "elevation": 8,
          "flexDirection": "row",
          "height": 56,
          "marginHorizontal": 16,
          "padding": 16,
          "shadowColor": "#0F172A",
          "shadowOffset": {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.15,
        }
      }
    >
      <Text
        style={
          [
            {
              "fontFamily": "PoppinsRegular",
              "fontSize": 14,
            },
            {
              "fontWeight": 400,
            },
            [
              {
                "color": "#FFFFFF",
              },
            ],
          ]
        }
      />
    </View>
  </View>,
]
`;
